src/core/client.ts

```diff
@@ -198,6 +198,24 @@ export default class Client {
     return true
   }
 
+  /**
+   * An async version of {@link notify} that resolves only after the notice has been reported to Honeybadger.
+   * Implemented using the {@link afterNotify} hook.
+   * Rejects if for any reason the report failed to be reported.
+   * Useful in serverless environments (AWS Lambda).
+   */
+  notifyAsync(noticeable: Noticeable, name: string | Partial<Notice> = undefined, extra: Partial<Notice> = undefined): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.afterNotify(err => {
```

Comment:
Doesn't using `this.afterNotify` mean that the handler is added to the client's permanent `afterNotify` handlers? What would happen if `notifyAsync` is called twice? Wouldn't the previous handler still fire, in addition to the new one?

Reply:
That's true. @subzero10 if the afterNotify handlers are exposed publicly (like process.listeners(event)`), we could ensure that the afterNotify function removes itself from the list when it's done.

Reply:
Correct. Good catch on noticing that @joshuap !
It now makes sense why I originally chose to use the `afterNotify` hook from the `notify` method.
@shalvah I'm not sure if it's a good idea to add the hook and then remove after when done. Consider this:
- `await Honeybadger.notifyAsync('message 1')`
- `await Honeybadger.notifyAsync('message 2')`
- Two `afterNotify` hooks are going to be registered on global `afterNotify` array and both will be called from the one that receives the response first, right? That would produce unexpected results.

It seems that the solution is:
- revert to the original `notify('message', { afterNotify })` implementation, but we have to make sure that this is handler is called last in the from the list of handlers.
@joshuap Is there a reason why we chose to run the handlers that come from the `notify()` method first, and not last?
```
export function runAfterNotifyHandlers(notice: Notice | null, handlers: AfterNotifyHandler[], error?: Error): boolean {
  if (notice && notice.afterNotify) {
    // why is this handler not pushed at the end of the array?
    handlers.unshift(notice.afterNotify)
  }

  for (let i = 0, len = handlers.length; i < len; i++) {
    handlers[i](error, notice)
  }
  return true
}
```

Reply:
Actually, I just wrote a test and it seems that we don't have to put the `notice.afterNotify` handler at the end of the array:
The current execution context will go all the way through the `afterNotifyHandler`s array, even if the promise is resolved with the first `afterNotify` handler. Resolving the promise simply puts the `then()` handler in a queue that will be invoked after the event loop takes back control. This will happen only after the current execution context is complete, which is when it goes through the array. Similar question [here](https://stackoverflow.com/questions/28896280/why-does-javascript-es6-promises-continue-execution-after-a-resolve).

With that said, I will go ahead and merge this PR (it's not in `master` branch, so we are pretty safe).

Reply:
I will let as is for now and will reconsider changing if I come across this again.